### PR TITLE
Expose `DistributedSampler` RNG seed argument

### LINKED ISSUE #1676: Finetuning dataloader does not respect shuffle seed
In `build_finetuning_dataloader`, either a [`StreamingDataset` is created based on the supplied configuration](https://github.com/mosaicml/llm-foundry/blob/bd113da06a9ca66890d628a9afc8b5c5ea7f02b1/llmfoundry/data/finetuning/dataloader.py#L255-L269) or a [`datasets.Dataset`](https://github.com/mosaicml/llm-foundry/blob/bd113da06a9ca66890d628a9afc8b5c5ea7f02b1/llmfoundry/data/finetuning/dataloader.py#L300-L312). For the `StreamingDataset`, its constructor arguments like `shuffle` and `shuffle_seed` are simply taken from the configuration with some minor filtering.
For the `datasets.Dataset`'s shuffling, afterwards, a [`DistributedSampler` is created](https://github.com/mosaicml/llm-foundry/blob/bd113da06a9ca66890d628a9afc8b5c5ea7f02b1/llmfoundry/data/finetuning/dataloader.py#L331-L339) which is passed the `shuffle` config value. However, it does not receive a potential `shuffle_seed`, meaning its shuffling is cannot be adjusted and will be deterministic across all runs.

(Below is old, incorrect text for the sake of @dakinggg's response making sense.)

~~In `build_finetuning_dataloader`, a [`StreamingDataset` is created based on the supplied configuration](https://github.com/mosaicml/llm-foundry/blob/bd113da06a9ca66890d628a9afc8b5c5ea7f02b1/llmfoundry/data/finetuning/dataloader.py#L300-L312). `StreamingDataset`'s constructor arguments like `shuffle` are simply taken from the configuration with some minor filtering. However, afterwards, a [`DistributedSampler` is created](https://github.com/mosaicml/llm-foundry/blob/bd113da06a9ca66890d628a9afc8b5c5ea7f02b1/llmfoundry/data/finetuning/dataloader.py#L334) which is again passed the `shuffle` config value. Since the `StreamingDataset` already received a value for `shuffle`, this is redundant.~~

~~Nothing is really wrong, but the second shuffling is simply not required.~~

### PR DESCRIPTION
**Pull request by janEbert**
Relies on https://github.com/mosaicml/composer/pull/3724.

Closes #1676.

### CODE REVIEW

**Review by mvpatel2000 — APPROVED** [review]
LGTM.

**Comment by mvpatel2000** [comment]
Note that tests will fail until Composer bump CC: @dakinggg  to track

**Comment by janEbert** [comment]
I guess we need to update the requirement in `setup.py` accordingly? Forgot about that.

**Comment by dakinggg** [comment]
Yeah unfortunately it won't work until we have the necessary code in a Composer release and then update Foundry. You can of course install from branches to unblock yourself as needed.

**Review by dakinggg — COMMENTED** [review]
The change itself lgtm though

**Comment by snarayan21** [comment]
Composer v0.28.0 has been released and dependency updated. Rerunning CI/CD
